lens-cube/src/main/java/org/apache/lens/cube/parse/StorageTableResolver.java

```diff
@@ -245,85 +232,109 @@ private void resolveDimStorageTablesAndPartitions(CubeQueryContext cubeql) throw
+   *
+   * TODO union : Do fourth point before 3.
    */
-
-
-  private void resolveFactStorageTableNames(CubeQueryContext cubeql) throws LensException {
-    Iterator<CandidateFact> i = cubeql.getCandidateFacts().iterator();
-    skipStorageCausesPerFact = new HashMap<>();
-    while (i.hasNext()) {
-      CubeFactTable fact = i.next().fact;
-
-      //TODO  this will be moved to CandidateTableResolver
-      if (fact.getUpdatePeriods().isEmpty()) {
-        cubeql.addFactPruningMsgs(fact, new CandidateTablePruneCause(CandidateTablePruneCode.MISSING_STORAGES));
-        i.remove();
+  private void resolveStorageTable(CubeQueryContext cubeql) throws LensException {
+    Iterator<Candidate> it = cubeql.getCandidates().iterator();
+    while (it.hasNext()) {
+      Candidate c = it.next();
+      assert (c instanceof StorageCandidate);
+      StorageCandidate sc = (StorageCandidate) c;
+      String storageTable = sc.getStorageName();
+      if (!isStorageSupportedOnDriver(storageTable)) {
+        log.info("Skipping storage: {} as it is not supported", storageTable);
+        cubeql.addStoragePruningMsg(sc, new CandidateTablePruneCause(CandidateTablePruneCode.UNSUPPORTED_STORAGE));
+        it.remove();
+        continue;
+      }
+      String str = conf.get(CubeQueryConfUtil.getValidStorageTablesKey(sc.getFact().getName()));
+      List<String> validFactStorageTables = StringUtils.isBlank(str)
+                                            ? null
+                                            : Arrays.asList(StringUtils.split(str.toLowerCase(), ","));
+      // Check if storagetable is in the list of valid storages.
+      if (validFactStorageTables != null && !validFactStorageTables.contains(storageTable)) {
+        log.info("Skipping storage table {} as it is not valid", storageTable);
+        cubeql.addStoragePruningMsg(sc, new CandidateTablePruneCause(CandidateTablePruneCode.INVALID_STORAGE));
+        it.remove();
         continue;
       }
 
-      Map<UpdatePeriod, Set<String>> storageTableMap = new TreeMap<UpdatePeriod, Set<String>>();
-      validStorageMap.put(fact, storageTableMap);
-      String str = conf.get(CubeQueryConfUtil.getValidStorageTablesKey(fact.getName()));
-      List<String> validFactStorageTables =
-        StringUtils.isBlank(str) ? null : Arrays.asList(StringUtils.split(str.toLowerCase(), ","));
-      Map<String, SkipStorageCause> skipStorageCauses = new HashMap<>();
-
-      for (Map.Entry<String, Set<UpdatePeriod>> entry : fact.getUpdatePeriods().entrySet()) {
-        String storage = entry.getKey();
-        // skip storages that are not supported
-        if (!isStorageSupported(storage)) {
-          log.info("Skipping storage: {} as it is not supported", storage);
-          skipStorageCauses.put(storage, new SkipStorageCause(SkipStorageCode.UNSUPPORTED));
-          continue;
+      boolean valid = false;
+      Set<CandidateTablePruneCause.CandidateTablePruneCode> codes = new HashSet<>();
```

Comment:
Can we call this pruningCauses? and also leave a comment that the cause can be different for different time ranges